com.unity.multiplayer.mlapi/Runtime/Core/NetworkBehaviour.cs

```diff
@@ -106,15 +126,31 @@ internal NetworkSerializer __beginSendClientRpc(uint rpcMethodId, ClientRpcParam
         {
             PooledNetworkWriter writer;
 
+            if (clientRpcParams.Send.UpdateStage == NetworkUpdateStage.Unset)
+            {
+                clientRpcParams.Send.UpdateStage = NetworkUpdateLoop.UpdateStage;
+
+                if (clientRpcParams.Send.UpdateStage == NetworkUpdateStage.Initialization)
+                {
+                    clientRpcParams.Send.UpdateStage = NetworkUpdateStage.EarlyUpdate;
+                }
+            }
+
+            if (clientRpcParams.Send.UpdateStage == NetworkUpdateStage.Initialization)
+            {
+                throw new NotSupportedException(
+                    $"{nameof(NetworkUpdateStage.Initialization)} cannot be used as a target for processing RPCs.");
+            }
+
             // This will start a new queue item entry and will then return the writer to the current frame's stream
-            var rpcQueueContainer = NetworkManager.RpcQueueContainer;
-            var isUsingBatching = rpcQueueContainer.IsUsingBatching();
+            var messageQueueContainer = NetworkManager.MessageQueueContainer;
+            var isUsingBatching = messageQueueContainer.IsUsingBatching();
             var transportChannel = rpcDelivery == RpcDelivery.Reliable ? NetworkChannel.ReliableRpc : NetworkChannel.UnreliableRpc;
 
-            ulong[] clientIds = clientRpcParams.Send.TargetClientIds ?? NetworkManager.ConnectedClientsList.Select(c => c.ClientId).ToArray();
+            ulong[] clientIds = clientRpcParams.Send.TargetClientIds ?? NetworkManager.ConnectedClientsIds;
```

Comment:
Oh boy, LINQ in a hot path. Great job getting rid of that!

Reply:
As much as I'd like to take credit for removing LINQ...

```C#
public ulong[] ConnectedClientsIds => ConnectedClientsList.Select(c => c.ClientId).ToArray();
```

I just noticed that was being repeated over and over and decided to functionalize it... or property-ize it, I guess.

Reply:
Yeah, the entire client id thing is an issue in this area and hard to get rid of completely without doing something like:
Create yet one more list that is nothing but ulong values of the connect clients.
We could change ConnectedClientsIds to a list and then in NetworkManager:
There are two places where ConnectedClientsList.RemoveAt is called, this could remove the ids from the new list.
The HandleApproval method is the only place that adds clients to the NetworkClient list and dictionary/look up table, and so that could populate the new list.
We would still have the cost of the .ToArray, but it would completely remove the LINQ stuff.

However, unfortunately that is most likely outside of the scope of this particular PR...but would be a great PR by itself!
:)